clang/lib/Sema/SemaBounds.cpp

```diff
@@ -4017,11 +4058,35 @@ namespace {
         << ObservedBounds << ObservedBounds->getSourceRange();
     }
 
+    // Returns the type of assignment E to show in the diagnostic messages:
+    // assignment (=), decrement (--) or increment (++).
+    Sema::BoundsDeclarationCheck GetBDCTypeFromExpr(Expr *E) const {
+      if (UnaryOperator *UO = dyn_cast<UnaryOperator>(E)) {
+        switch (UO->getOpcode()) {
+        case UnaryOperatorKind::UO_PreDec:
+        case UnaryOperatorKind::UO_PostDec:
+          return Sema::BoundsDeclarationCheck::BDC_Decrement;
+          break;
+        case UnaryOperatorKind::UO_PreInc:
+        case UnaryOperatorKind::UO_PostInc:
+          return Sema::BoundsDeclarationCheck::BDC_Increment;
+          break;
+        default:
+          break;
+        }
+      } else if (dyn_cast<BinaryOperator>(E)) {
+        // Must be an assignment or a compounds assignment, because E is
```

Comment:
If you're not using the result of `dyn_cast<BinaryOperator>(E)`, use `isa<BinaryOperator>(E)` instead.